functions/index.js

```diff
@@ -0,0 +1,208 @@
+const cheerio = require("cheerio");
+const cors = require("cors");
+const { Translate } = require("@google-cloud/translate");
+
+// require'd local modules
+const { proxyRequest } = require("./proxy");
+
+const corsHandler = cors({
+    // TODO: get deploy url
+    origin: ["localhost:3000"]
+});
+
+// variable initialization
+const TOO_MANY_SEGMENTS_MESSAGE = "Too many text segments";
+const projectId = "Genuis";
+// The target language
+const firstTarget = "ja";
+const secondTarget = "en";
+
+// Instantiates a translate client
+const translate = new Translate({
+    projectId
+});
+
+admin.initializeApp(functions.config().firebase);
+const db = admin.firestore();
+
+async function scrapeContent(path) {
+    const pageHtml = await axios
+        .get(`https://genius.com${path}`)
+        .then(({ data }) => {
+            return data;
+        });
+
+    const $ = cheerio.load(pageHtml);
+    const lyrics = $(".lyrics p");
+    const children = lyrics.contents();
+
+    const lyricsData = [];
+
+    children.map((_, element) => {
+        const el = $(element);
+        if (el[0].name === "br") return null;
+
+        const data = el[0].data;
+        if (data && data.startsWith("\n")) {
+            const newData = data.replace("\n", "").trim();
+            if (newData === "") return null;
+        }
+
+        const lyric = {
+            text: el.text().trim(),
+            referentId: el.data("id") || null
+        };
+
+        lyricsData.push(lyric);
+    });
+
+    return lyricsData;
+}
+
+async function translateSongLyrics(lyricsData) {
+    let lyricsDataCopy = lyricsData;
+    let retry = false;
+
+    do {
+        let lyricsText = lyricsDataCopy.map(({ text }) => text);
+
+        try {
+            // eslint-disable-next-line
+            let translationResults = await translate.translate(
+                lyricsText,
+                firstTarget
+            );
+            // Result of a translate call is always an array, the first item of which is the translated text/texts
+            let translatedLyrics = translationResults[0];
+
+            // eslint-disable-next-line
+            translationResults = await translate.translate(
+                translatedLyrics,
+                secondTarget
+            );
+            translatedLyrics = translationResults[0];
+
+            // eslint-disable-next-line
+            translatedLyrics.forEach((result, index) => {
+                lyricsDataCopy[index].text = result;
+            });
+            retry = false;
+        } catch (err) {
+            if (err.message === TOO_MANY_SEGMENTS_MESSAGE) {
+                lyricsDataCopy = lyricsDataCopy.slice(0, 99);
+                lyricsDataCopy.push({ text: "(song is too long)" });
+                retry = true;
+            } else {
+                throw err;
+            }
+        }
+    } while (retry);
+
+    return lyricsDataCopy;
+}
+
+exports.getWarpedSong = functions.https.onRequest(async (req, res) => {
```

Comment:
new endpoint `getWarpedSong` is given a songID and it will go and get that from firebase if it exists. If it doesn't it'll get it from Genius, translate it, store both the original and the warped version in firebase. Then return the warped version.